sdks/python/apache_beam/examples/snippets/snippets.py

```diff
@@ -110,80 +115,71 @@ def filter_words(unused_x):
-      reversed_words = words | ReverseWords()
-      # [END pipelines_constructing_applying]
+    # [START pipelines_constructing_writing]
+    filtered_words = reversed_words | 'FilterWords' >> beam.Filter(filter_words)
+    filtered_words | 'WriteMyFile' >> beam.io.WriteToText(
+        'gs://some/outputData.txt')
+    # [END pipelines_constructing_writing]
 
-      # [START pipelines_constructing_writing]
-      filtered_words = reversed_words | 'FilterWords' >> beam.Filter(
-          filter_words)
-      filtered_words | 'WriteMyFile' >> beam.io.WriteToText(
-          'gs://some/outputData.txt')
-      # [END pipelines_constructing_writing]
+    pipeline.visit(SnippetUtils.RenameFiles(renames))
 
-      p.visit(SnippetUtils.RenameFiles(renames))
 
-
-def model_pipelines(argv):
+def model_pipelines():
   """A wordcount snippet as a simple pipeline example."""
   # [START model_pipelines]
+  import argparse
   import re
+  import sys
 
   import apache_beam as beam
   from apache_beam.options.pipeline_options import PipelineOptions
 
-  class MyOptions(PipelineOptions):
-    @classmethod
-    def _add_argparse_args(cls, parser):
-      parser.add_argument(
-          '--input',
-          dest='input',
-          default='gs://dataflow-samples/shakespeare/kinglear'
-          '.txt',
-          help='Input file to process.')
-      parser.add_argument(
-          '--output',
-          dest='output',
-          required=True,
-          help='Output file to write results to.')
-
-  pipeline_options = PipelineOptions(argv)
-  my_options = pipeline_options.view_as(MyOptions)
-
-  with beam.Pipeline(options=pipeline_options) as p:
+  parser = argparse.ArgumentParser()
+  parser.add_argument(
+      '--input',
+      default='gs://dataflow-samples/shakespeare/kinglear.txt',
+      help='Input file to process.')
+  parser.add_argument(
+      '--output', required=True, help='Output file to write results to.')
+  args, beam_args = parser.parse_known_args(sys.argv)
 
+  beam_options = PipelineOptions(beam_args)
+  with beam.Pipeline(options=beam_options) as pipeline:
     (
-        p
-        | beam.io.ReadFromText(my_options.input)
+        pipeline
+        | beam.io.ReadFromText(args.input)
         | beam.FlatMap(lambda x: re.findall(r'[A-Za-z\']+', x))
         | beam.Map(lambda x: (x, 1))
         | beam.combiners.Count.PerKey()
-        | beam.io.WriteToText(my_options.output))
+        | beam.io.WriteToText(args.output))
   # [END model_pipelines]
 
 
-def model_pcollection(argv):
+def model_pcollection(output_path):
   """Creating a PCollection from data in local memory."""
   # [START model_pcollection]
+  import sys
+
   import apache_beam as beam
   from apache_beam.options.pipeline_options import PipelineOptions
 
-  # argv = None  # if None, uses sys.argv
-  pipeline_options = PipelineOptions(argv)
-  with beam.Pipeline(options=pipeline_options) as pipeline:
+  beam_options = PipelineOptions(sys.argv)
```

Comment:
This is equivalent to   
```
beam_options = PipelineOptions()
```

Reply:
Yes, I tried for a while with `mock` to get rid of `sys.argv`, but the only way that Beam actually recognizes the mocked versions are if I pass it explicitly here.

Please let me know if you have any suggestions on how to pass custom args for the tests, while keeping the snippet as simple as possible.

Reply:
Have you tried to mock `argparse._sys.argv` ? You need to apply the mock the module that actually uses the mocked object, in this case: https://github.com/python/cpython/blob/da5c808fb50d34bc2e180d9481706072f33025da/Lib/argparse.py#L90

Reply:
Thank you! I haven't tried that, I'll try that and update the rest of the snippets to do that.

Reply:
I tried patching `argparse._sys.argv` and it still doesn't recognize it.

Reply:
I also tried patching `apache_beam.examples.snippets.snippets.argparse._sys.argv`, and it still doesn't work.

Reply:
I wander if it's caching it somewhere or referencing a different memory address because it's a cpython module.

Reply:
This worked for me:
```
  def test_foo(self):
    import sys
    from unittest.mock import patch
    p = PipelineOptions()
    test_argv = ["prog", "--streaming"]
    with patch.object(sys, 'argv', test_argv):
      self.assertTrue(p.view_as(StandardOptions).streaming)
```

Reply:
Would something like this work for you?

Reply:
I've been trying to mock `argparse.ArgumentParser.parse_known_args` as well, but it's called multiple other times when building the PipelineOptions, so it's not working as expected. I still haven't found a good way to do this, if you have any other suggestions, please let me know.

Reply:
> Would something like this work for you?

For some reason, when I mock `sys.argv`, and I've spent several hours trying multiple different ways, when `args, beam_args = parser.parse_known_args()` runs in the sample, it doesn't get the new version unless I explicitly pass `sys.argv`.

Reply:
Hm, not sure. when you mock sys.argv, do you add one extra param to account for the first program name? Out of curiousity, I tried to remove argv in two random tests and didn't encounter issues (i ran tests via pytest locally and they seemed to pass.) example change: https://github.com/tvalentyn/beam/commit/6ebf753701dd3db5a0c502df18ebfe14098b5b0b#diff-0e735628d2670d783fa323985f4be5c86020c462f2bac6b514cc381037959c1eL674.

If there is a specific test where this doesn't work I can take a look, but don't have strong opinions here.  Leaving this up to you.

Reply:
and in this particular line you should definitely be able to use beam_options = PipelineOptions() without issues

Reply:
Thanks, it looks like I needed to add a first `unused` argument to the test_argv.

Reply:
that explains it. IIRC the first argv[0] is the path to the script itself, the rest is actual arguments.